usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateFrequencyStartTimeBeforeEndTime.java

```diff
@@ -0,0 +1,74 @@
+/*
+ *  Copyright (c) 2020. MobilityData IO.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.mobilitydata.gtfsvalidator.usecase;
+
+import org.apache.logging.log4j.Logger;
+import org.mobilitydata.gtfsvalidator.domain.entity.notice.error.FrequencyStartTimeAfterEndTimeNotice;
+import org.mobilitydata.gtfsvalidator.usecase.port.GtfsDataRepository;
+import org.mobilitydata.gtfsvalidator.usecase.port.ValidationResultRepository;
+import org.mobilitydata.gtfsvalidator.usecase.utils.TimeUtils;
+
+/**
+ * Use case to validate that `start_time` does not precede the `end_time`. This use case is triggered after completing
+ * the {@code GtfsDataRepository} provided in the constructor with {@code Frequency} entities.
+ */
+public class ValidateFrequencyStartTimeBeforeEndTime {
+    private final GtfsDataRepository dataRepo;
+    private final ValidationResultRepository resultRepo;
+    private final TimeUtils timeUtils;
+    private final Logger logger;
+
+    /**
+     * @param dataRepo   a repository storing the data of a GTFS dataset
+     * @param resultRepo a repository storing information about the validation process
+     * @param logger     a logger used to log information about the validation process
+     */
+    public ValidateFrequencyStartTimeBeforeEndTime(final GtfsDataRepository dataRepo,
+                                                   final ValidationResultRepository resultRepo,
+                                                   final TimeUtils timeUtils,
+                                                   final Logger logger) {
+        this.dataRepo = dataRepo;
+        this.resultRepo = resultRepo;
+        this.timeUtils = timeUtils;
+        this.logger = logger;
+    }
+
+    /**
+     * Use case execution method: checks if for each {@code Frequency} contained in {@link GtfsDataRepository} that
+     * `end_time` date does not precede the `start_time`. If this requirement is not met, a
+     * {@code StartTimeAfterEndTimeNotice} is added to the {@code ValidationResultRepo} provided in the
+     * constructor.
+     */
+    public void execute() {
+        logger.info("Validating rule 'E046 - `start_time` and `end_time` out of order");
+        dataRepo.getFrequencyAll().forEach((startTimeEndTime, frequency) -> {
+            // endTime and startTime are assumed to be not null at this stage. See Frequency builder.
```

Comment:
I'd suggest even stronger language:

```suggestion:
            // endTime and startTime cannot be not null at this stage. See Frequency builder.
```

Also, you really don't need to declare another two `int`s here - you could just reference `frequency.get*Time()` throughout the method and I think it's still very understandable.